### PR TITLE
Improve source mapping for .py and .pyi files

### PR DESCRIPTION
**Pull request by Aryan-SINGH-GIT**
Fixes issue #1850 ,

Solution:
  Added a mapping function that extracts a base name from protobuf-generated filenames and matches it to .proto files.
Changes:
* New function: map_python_protobuf_files() in scanpipe/pipes/d2d.py
* Extracts base names from *_pb2.py and *_pb3.py via regex
* Builds an index of .proto files by base name
* Creates CodebaseRelation with map_type="protobuf_mapping"
* integrated into: map_python step in DeployToDevelop pipeline
* Configuration update: Extended Python ecosystem in d2d_config.py to include .py and .pyi as matchable extensions

Tests: Added unit tests covering extraction, mapping, and edge cases
Testing:
* Unit tests pass
* Tested manually with the test data (6 mappings created)
image

Signed-off-by: Aryan Singh [aryansingh12oct2005@gmail.com](mailto:aryansingh12oct2005@gmail.com)

### CODE REVIEW

**Comment by Aryan-SINGH-GIT** [comment]
please review this PR @tdruez

File: scanpipe/pipes/d2d.py [hunk 1]
```diff
@@ -2372,3 +2372,50 @@ def map_python_pyx_to_binaries(project, logger=None):
...
+        base_name = _extract_protobuf_base_name(to_resource.name)
+        if base_name and base_name in proto_index:
+            from_resource = proto_index[base_name]
+            pipes.make_relation(
+                from_resource=from_resource,
+                to_resource=to_resource,
+                map_type="protobuf_mapping",
+                extra_data={"protobuf_base_name": base_name},
+            )
+            mapped_count += 1
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py]: The logic is fine, but you could add some line space to make the code a bit less dense to read:

```

    if not from_resources_count or not to_resources_count:
        return

    proto_index = {}
    for proto_resource in from_resources:
        base_name = proto_resource.name.replace(".proto", "")
        proto_index[base_name] = proto_resource

    mapped_count = 0
    for to_resource in to_resources:
        base_name = _extract_protobuf_base_name(to_resource.name)
        if base_name and base_name in proto_index:
            from_resource = proto_index[base_name]
            pipes.make_relation(
                from_resource=from_resource,
                to_resource=to_resource,
                map_type="protobuf_mapping",
                extra_data={"protobuf_base_name": base_name},
            )
            mapped_count += 1
```

File: scanpipe/pipes/d2d.py [hunk 2]
```diff
@@ -2372,3 +2372,50 @@ def map_python_pyx_to_binaries(project, logger=None):
...
+
+
+def _extract_protobuf_base_name(filename):
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py]: Any particular reason to make this function private?

File: scanpipe/pipes/d2d.py [hunk 3]
```diff
@@ -2372,3 +2372,50 @@ def map_python_pyx_to_binaries(project, logger=None):
...
+    """Extract the base name from a protobuf-generated filename."""
+    import re
+
+    name_without_ext = filename.rsplit(".", 1)[0]
+    protobuf_pattern = r"^(.+)_pb[23]$"
+    match = re.match(protobuf_pattern, name_without_ext)
+    if match:
+        return match.group(1)
+    return None
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py]: This line can be removed.

File: scanpipe/pipes/d2d.py [hunk 4]
```diff
@@ -2372,3 +2372,50 @@ def map_python_pyx_to_binaries(project, logger=None):
...
+                to_resource=to_resource,
+                map_type="protobuf_mapping",
+                extra_data={"protobuf_base_name": base_name},
+            )
+            mapped_count += 1
+
+
+def _extract_protobuf_base_name(filename):
+    """Extract the base name from a protobuf-generated filename."""
+    import re
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py]: Move the import at the module level.

File: scanpipe/pipes/d2d.py (hunk 2 above)
Comment (Aryan-SINGH-GIT) [reply scanpipe/pipes/d2d.py]: It was private because it was a small internal helper.
I will make it public since it’s a tested, reusable utility and matches the pattern of other public functions. I haven’t pushed the commit yet.

File: scanpipe/pipes/d2d.py (hunk 1 above)
Comment (Aryan-SINGH-GIT) [reply scanpipe/pipes/d2d.py]: done

File: scanpipe/pipes/d2d.py (hunk 3 above)
Comment (Aryan-SINGH-GIT) [reply scanpipe/pipes/d2d.py]: done

File: scanpipe/pipes/d2d.py (hunk 4 above)
Comment (Aryan-SINGH-GIT) [reply scanpipe/pipes/d2d.py]: done

**Comment by tdruez** [comment]
@Aryan-SINGH-GIT Looks great, thanks for your work on this one! 🚀
